keps/sig-cloud-provider/2699-add-webhook-hosting-to-ccm/README.md

```diff
@@ -0,0 +1,455 @@
+we already use to register new controllers. The existing sample shows this with
+a sample of registering the nodeipamcontroller which is not a normally installed
+controller in the cloud controller manager. In a similar way we will have a 
+sample of integrating a PVL mutating webhook into the sample CCM. We will also 
+have the system automatically detect if there are both controllers and webhooks
+registered in the binary. If both are registered it will automatically add 
+command line flags allowing webhooks and controller to be disabled. The 
+controller flag will default to the controllers being enabled. The webhook flag
+will default to the webhooks being disabled. We would also like to provide a 
+builder pattern for registering both the controller and webhook extensions.
+
+### User Stories (Optional)
+
+The users of this KEP are Cloud Providers. The intent is to make it easy for them
+to maintain the CCM controllers and webhooks across multiple versions. At the same
+time we are attempting to make it easy for the SIGs to make controllers or webhooks
+which can do what they know needs to be done and integrated into Cloud Provider 
+specific processes. We would like to do that in a way which makes merging upgrades
+relatively painless.
+
+#### Story 1
+
+Some Cloud Providers would prefer to keep controllers and webhooks in different 
+processses. They have concerns about attempting to run batch controllers in the 
+same process as webhooks which are "inline" and time sensitive. For these users
+it is easy to either build two different binaries or have the same binary act
+as two different binaries based on command line flags.
+
+#### Story 2
+
+For Cloud Providers who would like to keep things simple, it is easy to create
+a single processs which handles both controllers and webhooks.
+
+### Notes/Constraints/Caveats (Optional)
+
+
+### Risks and Mitigations
+
+Potentially there could be problems running webhooks and controller in the same
+process. Delays of 10 seconds or more can cause webhooks to fail. It is important 
+to understand than irrespective of failure mode on the webhook coniguration, 
+timeouts will always turn a webhook call into a FAIL. As such we are making it
+easy to easily turn the CCM into two processses to mitigate this. 
+
+We are actually changing the framework which generates the CCM and not the CCM
+itself. It has been pointed out that it is not the role of the controller
+manager to run webhooks. Controller managers should run controllers and webhooks
+are not controllers. As we are modifying the framework, we should consider this
+as we can create two processes. The CCM which just has controllers in it. We
+can also create a Cloud Webhook Manager. That is being left as homework for the 
+Cloud Provider. However the sample CCM which demonstrates how this will be done
+will have both in the same sample to make it easy.
+
+## Design Details
```

Comment:
For design details it might be nice to see example code of the hook that the cloud provider would use to register the webhook.

Reply:
Today we have things like app.NewCloudControllerManagerCommand(...).

We can continue to support that and add optional parameters to it.

However my primary thought was something like :
cmOptions, err := options.NewCloudManagerOptions()
if err != nil {
    klog.Fatalf("unable to initialize command options: %v", err)
}
fss := cliflag.NamedFlagSets{}
cloudManagerBuilder := app.NewCloudManagerBuilder("name")
cloudManagerBuilder.setOptions(cmOptions)
cloudManagerBuilder.setFlags(fss)
cloudManagerBuilder.registerWebhook(gvkList, handler)
cloudManagerBuilder.registerWebhook(gvkSecondList, secondHandler)
manager, err := cloudManagerBuilder(wait.NeverStop)
if err != nil {
    klog.Fatalf("unable to construct cloud manager: %v", err)
}
err := command.start()

Reply:
We could also then have a cloudControllerManagerBuilder with default controllers and additional methods like
cloudControllerManagerBuilder.registerController("nodeipamcontroller", handler)
and
cloudControllerManagerBuilder.deregisterController("servicecontroller")

Reply:
The builder pattern mostly makes the "enhance existing binary" easier but not the "generate a binary" part. Maybe it would be more neutral between new and existing binary to have `webhookBuilder` that eventually builds a `http.Handler` that can be either used in a new binary or CCM?

Reply:
Our existing controller manager framework, has made it easier to maintain something like the CCM. Once you have the CCM it is relatively easy to pull your dependencies (i.e. K/K) while maintaining the changes which make it your CCM. Unfortunately we've received feedback that the way you make changes (i.e. the extension hooks we provide) are not particularly intuitive. While we will maintain the existing mechanisms to support people who have already built their CCMs the idea here is to provide a more intuitive way to build your CCM.

Reply:
I believe it makes taking a "working" binary (the sample) and enhancing it to do what you want, easier. That is my goal here.